res/controllers/Traktor-Kontrol-S4-MK3-hid-scripts.js

```diff
@@ -0,0 +1,884 @@
+        partial_length = this.partial_packet.length;
+        for (var i = 0; i < length; i++) {
+            this.partial_packet[partial_length + i] = data[i];
+        }
+        this.controller.parsePacket(this.partial_packet,
+                                    partial_length + length);
+        // Clear out the partial packet
+        this.partial_packet = Object();
+        return;
+    }
+    
+    // This is the periodic report 0x03, can be parsed immediately.
+    if (length === 60 && data[0] === 3) {
+        this.controller.parsePacket(data, data.length);
+        return;
+    }
+    
+    this.error("Unhandled packet size: " + length + ", data[0] = " + data[0]);
+}
+
+// The logging functions, useful for debugging.
+
+TraktorS4MK3_alex.error = function(text) {
+    this.debug("ERROR: " + text)
+}
+
+TraktorS4MK3_alex.warning = function(text) {
+    this.debug("WARNING: " + text)
+}
+
+TraktorS4MK3_alex.info = function(text) {
+  HIDDebug("TraktorS4MK3: " + text)
+}
+
+TraktorS4MK3_alex.debug = function(text) {
+  HIDDebug("TraktorS4MK3: " + text)
+}
+
+TraktorS4MK3_alex.outOfOrderBitsWarning = function() {
+    this.warning("Received 15 bits. " +
+                 "Assuming this is the second half of the message. " +
+                 "But the first half has not been received yet, ignoring");
+}
+
+TraktorS4MK3_alex.noMappingError = function(s4ControlName) {
+    this.error("Unable to find mapping from Traktor S4 control " +
+               "'" + s4ControlName + "' to a Mixxx control")
+}
+
+TraktorS4MK3_alex.unknownGroupError = function(s4ControlName, mixxxControlName) {
+    this.error("Unable to identify the control group for the S4 control " +
+               "'" + s4ControlName + "' mapped to the Mixxx control " +
+               "'" + mixxxControlName + "'")
+}
+
+TraktorS4MK3_alex.unmappedDeckError = function(deck) {
+    this.error("Unable to map S4 deck '" + deck + "' to a Mixxx deck")
+}
+
+// NOTES on USB messages in Wireshark -----------------------------------------
+//
+// usb.darwin.endpoint_type == 1 is URB_ISOCHRONOUS
+// usb.darwin.endpoint_type == 2 is URB_BULK
+// usb.darwin.endpoint_type == 3 is URB_INTERRUPT
+//
+// usb.darwin.request_type == 0 is SUBMITTED
+// usb.darwin.request_type == 1 is COMPLETED
+//
+// usb.darwin.endpoint_address:
+//   0x00 - host
+//   0x01 - OUT, URB_ISOCHRONOUS, messages of ~22k bytes, ~700 KiB/s
+//   0x02 - OUT, URB_INTERRUPT, messages of 11, 41, 79 and 95 bytes (LEDs?)
+//   0x03 - OUT, URB_BULK, messages of 153k bytes, ~3 MiB/s
+//   0x81 - IN, URB_ISOCHRONOUS messages of ~2k bytes
+//   0x82 - IN, URB_ISOCHRONOUS messages of just 4 bytes
+//   0x83 - IN, messages 0x01, 0x02, and 0x03, URB_INTERRUPT mode
+//
+// Wireshark filter for the message3:
+//
+// usb.darwin.endpoint_address == 0x83 &&
+// usb.darwin.endpoint_type == 3 &&
+// ((usb.darwin.io_len == 60 && usb.darwin.request_type == 1)
+//  || (usb.darwin.io_len == 128 && usb.darwin.request_type == 0)))
+//
```

Comment:
Please move this info to the [Reverse Engineering wiki page](https://github.com/mixxxdj/mixxx/wiki/Reverse%20engineering).